src/evict/evict_lru.c

```diff
@@ -1385,9 +1385,15 @@ __evict_get_ref(
 	if (is_server && candidates > 1)
 		candidates /= 2;
 
-	/* Get the next page queued for eviction. */
+	/*
+	 * Get the next page queued for eviction: check both the WT_REF and the
+	 * page's flag: they're not set atomically, we can race with the server
+	 * setting them. Check the page's flag last, it's the element set last
+	 * and atomically.
+	 */
 	while ((evict = cache->evict_current) != NULL &&
-	    evict < cache->evict_queue + candidates && evict->ref != NULL) {
+	    evict < cache->evict_queue + candidates && evict->ref != NULL &&
+	    F_ISSET_ATOMIC(evict->ref->page, WT_PAGE_EVICT_LRU)) {
```

Comment:
I'm also surprised there is a race here - we add entries to the evict_queue linked list starting after the highest available candidate (see where ``slot = cache->evict_entries;`` in ``__evict_walk``). A candidate that is added to the queue in ``__evict_init_candidate`` should not be available to ``__evict_get_ref`` until after ``__evict_lru_walk`` updates the ``cache->evict_candidates`` field.